Add link crawler

barryvdh opened the pull request:
This adds a crawler to check all links on the homepage (not recursive/nested) for valid pages and returns the failed ones with status.

Small change to the crawler, to pass along the current URI, to generate correct crawl links.

phanan commented on app/Rules/WorkingLinks.php:
```diff
@@ -0,0 +1,118 @@
+use GuzzleHttp\Psr7\Request;
+use Psr\Http\Message\ResponseInterface;
+
+class WorkingLinks extends Rule
```
The rule name isn't very descriptive IMHO. I'd prefer something more verbose, for example, `NoBrokenImmediateLinks`, admittedly a big wordy.

barryvdh replied:
Changed it to `NoBrokenLinksOnPage`

phanan commented on app/Rules/WorkingLinks.php:
```diff
@@ -0,0 +1,118 @@
+    public function helpMessage()
+    {
+        return <<<'MSG'
+Make sure all links on your website are working.
```
Since we're only checking the immediate links on the checked page, maybe this message should be reworded correspondingly?

barryvdh replied:
Well, the help text is more of a general tip, right? but changed it :)

phanan commented:
Can you add a test as well?

barryvdh commented:
Not really sure, only when I can inject a mock client.

phanan commented:
I guess you can now ;)

barryvdh commented:
So something like this?

- Updated Client to be injected
- Added ClientServiceProvider to set defaults on the Client like UserAgent, so will affect all client that are injected with DI
- Try with HEAD request first, when fails, use GET (similar to your UrlHelper)
- Test with mock responses for success + failure
- Test to see if the actual requests match what we expect (2x HEAD, 1x GET) and relative links are correctly resolved.

phanan commented:
Beautiful ❤️
